### PR TITLE
Importer: endpoints and ui

### LINKED ISSUE #213: Importer endpoint and UI interface
Backend: create a login/permission protected endpoint which expects a payload

`{
'provider': ..., 
'mode': ...
}`

And allows to upload a file. File should be validated as JSON or XML (check mimetype allowed in the REST endpoints). The endpoint should have a possibility to login with a token too (it should work out of the box on a invenio user account - you need to create one dedicated, needs only a test)

Providers can come from a config variable or from a vocabulary. Let's discuss briefly whats better.

UI:
* user should be able to choose a provider from the list.
* user should have a very simple interface to upload the file. File should be XML type and validates as such. Write it in a way to leave a possibility to upload JSON files in the future
* mode might be 'modify/create' or 'delete'. Ask the user for confirmation on the delete action

* on a success action (might be long, the status should be checked every 2 minutes) the user should be able to see the imported documents and items/eitems created (you can simulate it if the backend part not done yet)

The goal here is to have an entry point which allows to upload an XML file with records from providers, and it returns the links to imported records 

Bonus task (if we have time - otherwise please ticketize it)

- [ ] reverting the last action - if the librarian reviews the action and thinks that the whole thing was a mistake, we should be able to revert it. Caution: it requires accessing the record versions table, so lets evaluate it together

### PR DESCRIPTION
**Pull request by jrcastro2**
* closes #213
* ~depends on inveniosoftware/react-invenio-app-ils#308~ :heavy_check_mark: 

This new page allows us to import documents slecting the provider, the mode and uploading the file.

- All the fields are required. 
- The file estension must be `.xml`.
- If we select `delete` as mode, a confirmation modal will appear.
- While processing data (importing records), if we try to close/reload the results page we will get an alert.
- Each 30 seconds we will fetch the data (this can be changed).

Some screenshots:

### Importer form
image

### Form validation (frontend)
image

### Display of backend errrors on POST
image

### Confirmation modal on delete
image

### Loading
image

### Alert while loading on close/reload of tab
image

### Import failed
image

### Results page
image

## Invenio frontend

image

image

image

### CODE REVIEW

**Comment by jrcastro2** [comment]
Note for reviewers: Please just review last commit (the UI changes), the backend changes can be checked [here](https://github.com/CERNDocumentServer/cds-ils/pull/236).

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                      active={activeIndex === index}
+                      index={index}
+                      onClick={this.handleClick}
+                      color="red"
+                    >
+                      <Icon name="dropdown" />
+                      {!_isEmpty(document) ? (
+                        <Link
+                          to={BackOfficeRoutes.documentDetailsFor(document.pid)}
+                          target="_blank"
```
Comment (FlorianCassayre) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:159]: 👍  nice detail

File: cds_ils/patrons/permissions.py [hunk 2]
```diff
@@ -25,9 +25,17 @@ def retrieve_patron_loans_permission(*args, **kwargs):
     )
 
 
+def backoffice_permission(*args, **kwargs):
+    """Return permission to allow only librarians and admins."""
+    return Permission(backoffice_access_action)
```
Comment (FlorianCassayre) [inline cds_ils/patrons/permissions.py:30]: Hm, why do we need to redefine `backoffice_permission`? (there is one in ILS already)

**Review by FlorianCassayre — COMMENTED** [review]
Good job, very nice UI!
A couple of minor remarks:
* the first progress update is important because it contains valuable informations (how many entries have to be processed) but it takes some time to receive it (30 seconds currently)
* the top border appears to be missing:
image

File: cds_ils/patrons/permissions.py (hunk 2 above)
Comment (jrcastro2) [reply cds_ils/patrons/permissions.py:30]: Yes, you are right this code is lefotver from a bad rebase, thanks!

Comment (FlorianCassayre) [reply cds_ils/patrons/permissions.py:30]: Okay! I've implemented the importer permissions in that file already but maybe it's not the best place. In any case this will be done in the other PR.

File: cds_ils/importer/views.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                else:
+                    reports.append({
+                        "index": entry.entry_index,
+                        "success": False,
+                        "message": entry.error,
+                    })
+            obj["reports"] = reports
+            return obj
+        else:
+            raise Exception("Task not found")  # TODO
```
Comment (ntarocco) [inline cds_ils/importer/views.py:113]:
```suggestion
            return 404
```

If you want to keep it cleaner, maybe return at the top:
```
if not log:
    return 404
```
or extract:
```
if log:
    return get_report(log)
else:
    return 404
```

File: cds_ils/importer/views.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                t.start()
+
+                return (json.dumps({"id": log.id}),
+                        200,
+                        {'ContentType': 'application/json'})
+            else:
+                flash('That file extension is not allowed')
+                raise RequestError('That file extension is not allowed', 400)
+        else:
+            raise RequestError('Missing file', 400)
```
Comment (ntarocco) [inline cds_ils/importer/views.py:167]: I would replace all `raise`, in a view/request context, with `abort`.

File: cds_ils/importer/views.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            obj["reports"] = reports
+            return obj
+        else:
+            raise Exception("Task not found")  # TODO
+
+    @blueprint.route('/importer', methods=['POST'])
+    @need_permissions("document-importer")
+    def importer():
+
+        @copy_current_request_context
```
Comment (ntarocco) [inline cds_ils/importer/views.py:119]: why is this needed?

File: cds_ils/importer/views.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                raise RequestError('Missing mode', 400)
+
+            if allowed_files(file.filename):
+                original_filename = file.filename
+                file.filename = rename_file(file.filename)
+                filename = secure_filename(file.filename)
+                file.save(os.path.join(app.config["IMPORTER_UPLOADS_PATH"],
+                                       filename))
+
+                # TODO move this to API
```
Comment (ntarocco) [inline cds_ils/importer/views.py:145]: TODO warning

File: cds_ils/importer/views.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                log = ImporterTaskLog.create(dict(
+                    agent=ImporterAgent.USER,
+                    provider=provider,
+                    source_type=source_type,
+                    mode=ImporterMode.CREATE if mode == "create"
+                    else ImporterMode.UPDATE,
+                    original_filename=original_filename,
+                ))
+
+                t = Thread(target=process_data,
```
Comment (ntarocco) [inline cds_ils/importer/views.py:156]: why a new thread in a request context? What is the advantage?

File: cds_ils/importer/views.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+                        "message": entry.error,
+                    })
+            obj["reports"] = reports
+            return obj
+        else:
+            raise Exception("Task not found")  # TODO
+
+    @blueprint.route('/importer', methods=['POST'])
+    @need_permissions("document-importer")
+    def importer():
```
Comment (ntarocco) [inline cds_ils/importer/views.py:117]: can we add a max size validation?

Comment (ntarocco) [inline cds_ils/importer/views.py:117]: did you try to upload a `bla.xml` which contain a non valid XML? Everything good?

File: cds_ils/importer/views.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+                      + file.filename) as f:
+                import_from_xml(log_id, f, source_type, provider)
+
+        if request.files:
+            file = request.files["file"]
+            provider = request.form.get("provider", None)
+            mode = request.form.get("mode", None)
+
+            if not provider:
+                flash('Missing provider')
```
Comment (ntarocco) [inline cds_ils/importer/views.py:131]: isn't `flash` used only in Jinja templates? Why is it needed?

File: cds_ils/importer/views.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+        @copy_current_request_context
+        def process_data(log_id, file, source_type, provider):
+            with open(app.config["IMPORTER_UPLOADS_PATH"] + "/"
+                      + file.filename) as f:
+                import_from_xml(log_id, f, source_type, provider)
+
+        if request.files:
+            file = request.files["file"]
+            provider = request.form.get("provider", None)
```
Comment (ntarocco) [inline cds_ils/importer/views.py:127]: Please prefer `webargs` instead, you have args validation and defaults out of the box.

File: cds_ils/importer/views.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+                raise RequestError('Missing provider', 400)
+
+            if not mode:
+                flash('Missing mode')
+                raise RequestError('Missing mode', 400)
+
+            if allowed_files(file.filename):
+                original_filename = file.filename
+                file.filename = rename_file(file.filename)
+                filename = secure_filename(file.filename)
```
Comment (ntarocco) [inline cds_ils/importer/views.py:141]: renaming 2 times? Why is that?

File: cds_ils/ldap/tasks.py [hunk 12]
```diff
@@ -28,5 +29,6 @@ def synchronize_users_task():
         result = update_users()
         log.set_succeeded(*result)
     except Exception as e:
+        db.session.rollback()
```
Comment (ntarocco) [inline cds_ils/ldap/tasks.py:32]: 👍 great! Were you able to validate that it actually fixes the issue?

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+      activeIndex: -1,
+      importCompleted: false,
+      data: null,
+      intervalId: null,
+      isLoading: true,
+    };
+  }
+
+  componentDidMount() {
+    //TODO: Define how much we wait until the next call (currently 30 sec)
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:27]: TODO warning. Why not using a config instead?
30 secs looks a lot to me... Then I would expect to see something like: checking in 30... 29... etc...

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    const idVar = setInterval(() => this.checkForData(idVar), 30000);
+    // eslint-disable-next-line react/no-did-mount-set-state
+    this.setState({ intervalId: idVar });
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:30]: No need for React state here:
```suggestion
    this.intervalId = idVar;
```

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+    this.checkForData(idVar);
+    window.addEventListener('beforeunload', this.onCloseWarning);
+  }
+
+  componentWillUnmount = () => {
+    const { intervalId } = this.state;
+    window.removeEventListener('beforeunload', this.onCloseWarning);
+    clearInterval(intervalId);
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:38]: Safer to check if defined:
```suggestion
    this.intervalId && clearInterval(this.intervalId);
```

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    const idVar = setInterval(() => this.checkForData(idVar), 30000);
+    // eslint-disable-next-line react/no-did-mount-set-state
+    this.setState({ intervalId: idVar });
+    this.checkForData(idVar);
+    window.addEventListener('beforeunload', this.onCloseWarning);
+  }
+
+  componentWillUnmount = () => {
+    const { intervalId } = this.state;
+    window.removeEventListener('beforeunload', this.onCloseWarning);
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:37]: Does it failed if not defined for some reasons? Otherwise:
```
this.onCloseWarning && window.removeEventListener('beforeunload', this.onCloseWarning);
```

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    clearInterval(intervalId);
+  };
+
+  onCloseWarning = e => {
+    const { isLoading } = this.state;
+    if (isLoading) {
+      var confirmationMessage =
+        'Importing in progress. Are you sure you want to leave?';
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:45]: What is happening if I leave? Everything lost? Then better to warn the user:
`Importing in progress. If you leave the page, the import will be cancelled. Are you sure?`

**Review by ntarocco — COMMENTED** [review]
One of the main comment is that I have the feeling that if it is a long task, we should this in a celery task so that if I leave the web page, the process is not lost.
There are a lot of things that can go wrong just waiting on a webpage: loosing connection, closing the laptop lid, the battery of the laptop dies... :) this is one of the biggest issue with have in CDS Videos.
Maybe to be discussed in IRL.

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+    }
+  };
+
+  checkForData = async idVar => {
+    const { importCompleted } = this.state;
+    const { taskId } = this.props;
+
+    if (!importCompleted) {
+      // eslint-disable-next-line no-useless-concat
+      const response = await http.get(`${importerURL}` + '/' + `${taskId}`);
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:58]: I would have maybe created a nice endpoint in the `api.py` as for the others, instead of having it inside a component.

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+          </>
+        ) : data.state === 'SUCCEEDED' ? (
+          <>
+            <Icon name="check" />
+            Documents imported succesfully.
+          </>
+        ) : (
+          <>
+            <Icon name="close" />
+            Documents import failed.
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:114]: It can be also Series and EItems? If yes, I would use Literature everywhere.

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+                    ' documents out of ' +
+                    data.total_entries +
+                    '.'}
+                </p>
+              ) : (
+                <p>{'Processed ' + data.reports.length + ' documents.'}</p>
+              )
+            ) : null}
+            {!_isEmpty(data.reports) ? (
+              <Accordion className="importer" styled fluid>
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/ImportedDocuments.js:146]: can we refactor and extract in smaller cmps?

File: ui/src/overridden/backoffice/Importer/Importer.js [hunk 21]
```diff
@@ -0,0 +1,253 @@
...
+
+  postData = async formData => {
+    const importerURL = '/importer';
+    const headers = {
+      headers: {
+        'Content-Type': 'multipart/form-data',
+      },
+    };
+    try {
+      const response = await http.post(`${importerURL}`, formData, headers);
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/Importer.js:53]: same here for the endpoint, to have a module in `api` to have all endpoints together.

File: ui/src/overridden/backoffice/Importer/Importer.js [hunk 22]
```diff
@@ -0,0 +1,253 @@
...
+      this.setState({
+        error: error,
+      });
+    }
+  };
+
+  handleSubmit = async () => {
+    const { provider, mode, file } = this.state;
+    if (!_isEmpty(provider) && !_isEmpty(mode) && file) {
+      var formData = new FormData();
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/Importer.js:72]: we are using a React upload component to upload files for eitems, why not re-using that one and get rid of custom code?

File: ui/src/overridden/backoffice/Importer/Importer.js [hunk 23]
```diff
@@ -0,0 +1,253 @@
...
+      </Form>
+    );
+  };
+
+  render() {
+    const { postDone, error } = this.state;
+    return (
+      <Container id="importer" className="spaced">
+        <>
+          <Header as="h1">Documents Importer</Header>
```
Comment (ntarocco) [inline ui/src/overridden/backoffice/Importer/Importer.js:241]:
```suggestion
          <Header as="h1">Literatures Importer</Header>
```

File: cds_ils/importer/views.py (hunk 5 above)
Comment (jrcastro2) [reply cds_ils/importer/views.py:119]: Needed for the thread to "work" inside the view function.

File: cds_ils/importer/views.py (hunk 7 above)
Comment (jrcastro2) [reply cds_ils/importer/views.py:156]: It's used to launch the importing process and to be able to return a 200 (if everything ok) at the same time. Otherwise we would be waiting for the response

File: cds_ils/importer/views.py (hunk 8 above)
Comment (jrcastro2) [reply cds_ils/importer/views.py:117]: Yes, an error warning you that the import failed will appear.

File: cds_ils/importer/views.py (hunk 9 above)
Comment (jrcastro2) [reply cds_ils/importer/views.py:131]: Yes, will remove it, it's leftover code, sorry and thanks!

File: cds_ils/importer/views.py (hunk 11 above)
Comment (jrcastro2) [reply cds_ils/importer/views.py:141]: You are right, the second one is no longer needed since we have unique names.

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js (hunk 16 above)
Comment (jrcastro2) [reply ui/src/overridden/backoffice/Importer/ImportedDocuments.js:37]: It doesn't fail but I will add it, just in case

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js (hunk 17 above)
Comment (jrcastro2) [reply ui/src/overridden/backoffice/Importer/ImportedDocuments.js:45]: No, if you leave you will not be able to keep track of the import, but it will keep running until it succeeds or fails.

File: ui/src/overridden/backoffice/Importer/ImportedDocuments.js (hunk 19 above)
Comment (jrcastro2) [reply ui/src/overridden/backoffice/Importer/ImportedDocuments.js:114]: If I am not wrong we import documents, but we also create/update/delete eitems related to them, so yes we can use Literature

**Comment by FlorianCassayre** [comment]
Some of the comments are for the other PR, I'll take care of them.

File: ui/src/overridden/backoffice/Importer/Importer.js (hunk 22 above)
Comment (jrcastro2) [reply ui/src/overridden/backoffice/Importer/Importer.js:67]: Hmm, if you mean this component I did not use it because it seems to specific for the eitems, it needs an eitem as a prop, it launches the call on selecting the file, while we only want to store it and send it along with the rest of the data from the form. Maybe I can create a more generic component re-usable by both use cases

File: cds_ils/importer/views.py (hunk 8 above)
Comment (jrcastro2) [reply cds_ils/importer/views.py]: Added, but maybe the current max size is too much ? (100 MiB) (see [here](https://github.com/CERNDocumentServer/cds-ils/blob/master/cds_ils/config.py#L186))

File: cds_ils/ldap/tasks.py (hunk 12 above)
Comment (FlorianCassayre) [reply cds_ils/ldap/tasks.py:32]: I tested it this morning and it was in fact the cause of the issue of the tasks being left "RUNNING". We had a test that covered failures, but database errors are different in the sense that they have side-effects (and we must rollback the changes when one occurs, before committing further changes). The errors are now correctly reported in the database.

File: cds_ils/importer/views.py (hunk 8 above)
Comment (FlorianCassayre) [reply cds_ils/importer/views.py]: Hmm maybe, the main danger is that we are running it on the webserver node, so we must be careful not to load too much data in memory.

File: cds_ils/importer/api.py [hunk 24]
```diff
@@ -48,48 +48,39 @@ def import_record(data, provider, source_type=None, eager=False):
...
-    """Load xml file."""
-    for idx, source in enumerate(sources, 1):
-        click.echo(
-            "({}/{}) Importing documents in {}...".format(
-                idx, len(sources), source.name
-            )
-        )
+def import_from_xml(log_id, source, source_type, provider):
+    """Load a singular xml file."""
+    log = ImporterTaskLog.query.filter_by(id=log_id).first()
```
[the remark this answers is not in the document]
Comment (FlorianCassayre) [reply cds_ils/importer/api.py:53]: I was encountering issues when changing context if I passed the database object. I'll see what I can do.

File: cds_ils/config.py [hunk 25]
```diff
@@ -631,3 +631,7 @@ def _parse_env_bool(var_name, default=None):
     "ebl": {"priority": 3, "agency_code": "MiAaPQ"},
     "safari": {"priority": 4, "agency_code": "CaSebORM"},
 }
+
+CDS_ILS_IMPORTER_UPLOADS_PATH = "/tmp"
+
+CDS_ILS_IMPORTER_FILE_EXTENSIONS_ALLOWED = ["xml"]
```
Comment (ntarocco) [inline cds_ils/config.py:637]: Very minir: no `.` in the end? Otherwise I could probably upload something `.axml` :)

File: cds_ils/importer/views.py [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+        """Renames filename with an unique name."""
+        unique_filename = uuid.uuid4().hex
+        ext = filename.rsplit(".", 1)[1]
+        return unique_filename + "." + ext
+
+    @blueprint.errorhandler(413)
+    def payload_too_large(error):
+        return {"message": "The file is too large", "status": 413}, 413
+
+    @blueprint.route("/importer/check/<int:log_id>", methods=["GET"])
```
Comment (ntarocco) [inline cds_ils/importer/views.py:47]: why not simply:
- `/importer/tasks/` for the list
- `/importer/tasks/<job_id>` for the details
(probably to be added to the new GH issue)

File: cds_ils/importer/views.py [hunk 27]
```diff
@@ -0,0 +1,171 @@
...
+    @need_permissions("document-importer")
+    def check(log_id):
+        return check_next(log_id, 0)
+
+    @blueprint.route("/importer/check/<int:log_id>/next/<int:next_entry>",
+                     methods=["GET"])
+    @need_permissions("document-importer")
```
Comment (ntarocco) [inline cds_ils/importer/views.py:54]: can you explain why the `next` param? Is it like a range? Why is it needed?

File: cds_ils/importer/views.py [hunk 28]
```diff
@@ -0,0 +1,171 @@
...
+                        "index": entry.entry_index,
+                        "success": False,
+                        "message": entry.error,
+                    })
+            obj["reports"] = reports
+            return obj
+        else:
+            return 404
+
+    @blueprint.route("/importer", methods=["POST"])
```
Comment (ntarocco) [inline cds_ils/importer/views.py:107]:
```suggestion
    @blueprint.route("/importer/tasks", methods=["POST"])
```
to be added to the ticket?

File: cds_ils/importer/views.py [hunk 29]
```diff
@@ -0,0 +1,171 @@
...
+                file.save(source_path)
+
+                source_type = "marcxml"
+                log_id = create_import_task(source_path,
+                                            original_filename,
+                                            source_type,
+                                            provider, mode)
+
+                return (json.dumps({"id": log_id}),
+                        200,
```
Comment (ntarocco) [inline cds_ils/importer/views.py:164]:
```suggestion
                        201,
```

**Review by ntarocco — APPROVED** [review]

File: cds_ils/config.py (hunk 25 above)
Comment (FlorianCassayre) [reply cds_ils/config.py:637]: Nice catch, you're totally right.

File: cds_ils/importer/views.py (hunk 27 above)
Comment (FlorianCassayre) [reply cds_ils/importer/views.py:54]: Should have been documented; the idea is to avoid fetching all the entries each time because they don't change, and instead only ask for the ones that we don't have. Reason is that the response payload size grows linearly otherwise (and fast!).
